pkg/enforcement/user_permissions.go

```diff
@@ -74,50 +74,104 @@ func (e *PermitEnforcer) parseUserPermissionsResponse(res *http.Response) (UserP
+	ResourceTypes []string          `json:"resource_types,omitempty"`
+	Resources     []string          `json:"resources,omitempty"`
+	Context       map[string]string `json:"context,omitempty"`
 }
 
-func NewGetUserPermissionsRequest(user User, tenants []string) *GetUserPermissionsRequest {
-	return &GetUserPermissionsRequest{
-		User:    user,
-		Tenants: tenants,
-		Context: nil,
+// UserPermissionOption is a function that modifies GetUserPermissionsRequest
+type UserPermissionOption func(*GetUserPermissionsRequest)
+
+// WithTenants adds tenant filters to the user permissions request
+func WithTenants(tenants []string) UserPermissionOption {
+	return func(r *GetUserPermissionsRequest) {
+		r.Tenants = tenants
 	}
 }
 
-func newJsonGetUserPermissionsRequest(opaUrl string, user User, tenants []string) ([]byte, error) {
-	getUserPermissionsReq := NewGetUserPermissionsRequest(user, tenants)
-	var genericCheckReq interface{} = getUserPermissionsReq
-	if opaUrl != "" {
-		genericCheckReq = &struct {
-			Input *GetUserPermissionsRequest `json:"input"`
-		}{getUserPermissionsReq}
+// WithResourceTypes adds resource type filters to the user permissions request
+func WithResourceTypes(resourceTypes []string) UserPermissionOption {
+	return func(r *GetUserPermissionsRequest) {
+		r.ResourceTypes = resourceTypes
 	}
-	jsonCheckReq, err := json.Marshal(genericCheckReq)
-	if err != nil {
-		return nil, err
+}
+
+// WithResources adds resource instance filters to the user permissions request
+func WithResources(resources []string) UserPermissionOption {
+	return func(r *GetUserPermissionsRequest) {
+		r.Resources = resources
 	}
-	return jsonCheckReq, nil
 }
 
-func (e *PermitEnforcer) GetUserPermissions(user User, tenants ...string) (UserPermissions, error) {
-	reqAuthValue := "Bearer " + e.config.GetToken()
+// WithContext adds context data to the user permissions request
+func WithContext(context map[string]interface{}) UserPermissionOption {
+	return func(r *GetUserPermissionsRequest) {
+		// Convert the map[string]interface{} to map[string]string for compatibility
+		if context != nil {
+			contextMap := make(map[string]string)
+			for k, v := range context {
+				// Convert value to string if it's not already
+				switch val := v.(type) {
+				case string:
+					contextMap[k] = val
+				case bool:
+					if val {
+						contextMap[k] = "true"
+					} else {
+						contextMap[k] = "false"
+					}
+				default:
+					// Try to marshal other types to JSON
+					jsonBytes, err := json.Marshal(v)
+					if err == nil {
+						contextMap[k] = string(jsonBytes)
+					}
+				}
+			}
+			r.Context = contextMap
+		}
+	}
+}
 
-	jsonCheckReq, err := newJsonGetUserPermissionsRequest(e.config.GetOpaUrl(), user, tenants)
+// GetUserPermissions returns permissions a user has, optionally filtered by tenants, resource types, resources, and with custom context
+func (e *PermitEnforcer) GetUserPermissions(user User, opts ...UserPermissionOption) (UserPermissions, error) {
```

Comment:
You are breaking the function signature here - the original one had a required tenant argument which is absent right now.
Also the opts should be optional.